### PR TITLE
Additional validation for all draw methods

### PR DESCRIPTION
**Pull request by toji**
Based on conversations with @kainino0x it sounds like there's some reshaping in progress for the pipeline creation (ie: #767) which may necessitate changes and additions here, but this should be a good place to start.

Required me to reformat `createPipeline()` in order to make the descriptor accessible as an internal slot, which also led my to realized that algorithm had the wrong return type. Oops!

***
<a href="https://pr-preview.s3.amazonaws.com/toji/gpuweb/pull/1010.html" title="Last updated on Aug 19, 2020, 9:32 PM UTC (48ec2ec)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1010/9b3a658...toji:48ec2ec.html" title="Last updated on Aug 19, 2020, 9:32 PM UTC (48ec2ec)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -5475,7 +5500,14 @@ enum GPUStoreOp {
 
             **Returns:** {{undefined}}
 
-            Issue: Describe {{GPURenderEncoderBase/drawIndexed()}} algorithm steps.
+            Issue the following steps on the [=Device timeline=] of |this|.{{GPUObjectBase/[[device]]}}:
+            <div class=device-timeline>
+                If any of the following conditions are unsatisfied, make |this| [=invalid=] and stop.
+                <div class=validusage>
+                    - It is [$valid to draw$] with |this|.
+                    - |this|.{{GPURenderEncoderBase/[[index_buffer]]}} is not `null`.
```
Comment (kainino0x) [inline spec/index.bs:5508]: Indexed draw also needs to check that the indexFormat of the pipeline, if set, is equal to the indexFormat of the most recent setIndexBuffer. (bleh, extra draw time validation 🙁 )

Comment (toji) [reply spec/index.bs]: Done. Added a new internal slot to the pipeline (`[[strip_index_format]]`) to keep things from getting too wordy.

File: spec/index.bs [hunk 2]
```diff
@@ -3335,6 +3335,20 @@ GPURenderPipeline includes GPUObjectBase;
...
+<dl dfn-type=attribute dfn-for=GPURenderPipeline>
+    : <dfn>\[[descriptor]]</dfn>, of type {{GPURenderPipelineDescriptor}}
+    ::
+        The {{GPURenderPipelineDescriptor}} describing this pipeline.
+
+        All optional fields of {{GPURenderPipelineDescriptor}} are defined.
+
+    : <dfn>\[[strip_index_format]]</dfn>, of type {{GPUIndexFormat}}?
+    ::
+        The format index data this pipeline requires, initially `null`.
```
Comment (Kangz) [inline spec/index.bs:3349]: nit: `undefined` instead of null maybe to match the descriptor?

File: spec/index.bs [hunk 3]
```diff
@@ -5562,6 +5603,43 @@ enum GPUStoreOp {
...
+                - Let |bindGroup| be |encoder|.{{GPUProgrammablePassEncoder/[[bind_groups]]}}[|index|].
+                - |bindGroup| must not be `null`.
+                - |bindGroup|.{{GPUBindGroup/[[layout]]}} must be [=group-equivalent=] with |bindGroupLayout|.
+
+            - Let |pipelineDescriptor| be |encoder|.{{GPURenderEncoderBase/[[pipeline]]}}.{{GPURenderPipeline/[[descriptor]]}}.
+            - For each {{GPUIndex32}} |slot| `0` to
+                |pipelineDescriptor|.{{GPURenderPipelineDescriptor/vertexState}}.{{GPUVertexStateDescriptor/vertexBuffers}}.length:
+                - |encoder|.{{GPURenderEncoderBase/[[vertex_buffers]]}}[|slot|] must not be `null`.
+        </div>
+    1. Return `true`.
```
Comment (Kangz) [inline spec/index.bs:5625]: Can we have an issue to check for the buffer binding sizes when `minBufferBindingSize` isn't set?

Also this part + having a pipeline could be refactored to be common to all programmable passes (compute and render).

**Review by Kangz — APPROVED** [review]

File: spec/index.bs [hunk 4]
```diff
@@ -3444,33 +3458,46 @@ Issue: link to the semantics of SV_SampleIndex and SV_Coverage in WGSL spec.
...
-            Then:
-                1. Generate a {{GPUValidationError}} in the current scope with appropriate error message.
-                1. Create a new [=invalid=] {{GPURenderPipeline}} and return the result.
+                        Then:
+                            1. Generate a {{GPUValidationError}} in the current scope with appropriate
+                                error message.
+                            1. Make |pipeline| [=invalid=].
+
+                    1. Set |pipeline|.{{GPURenderPipeline/[[descriptor]]}} to |descriptor|.
+                    1. If |descriptor|.{{GPURenderPipelineDescriptor/primitiveTopology}} is
```
Comment (Kangz) [inline spec/index.bs:3494]: I think we should have validation that `indexFormat` is set IFF the primitive topology is a strip.

Comment (toji) [reply spec/index.bs:3511]: Done!

File: spec/index.bs (hunk 3 above)
Comment (toji) [reply spec/index.bs]: Added the issue and refactored the bind group validation so that it applies to any type of encoder.

File: spec/index.bs (hunk 2 above)
Comment (toji) [reply spec/index.bs]: Done!

**Comment by toji** [comment]
Thanks @Kangz! Addressed all your feedback.

File: spec/index.bs [hunk 5]
```diff
@@ -5587,6 +5642,41 @@ enum GPUStoreOp {
...
+    run the following steps:
+
+    If any of the following conditions are unsatisfied, return `false`:
+        <div class=validusage>
+            - [$Validate encoder bind groups$](|encoder|, |encoder|.{{GPURenderEncoderBase/[[pipeline]]}})
+                must be `true`.
+
+            - Let |pipelineDescriptor| be |encoder|.{{GPURenderEncoderBase/[[pipeline]]}}.{{GPURenderPipeline/[[descriptor]]}}.
+            - For each {{GPUIndex32}} |slot| `0` to
+                |pipelineDescriptor|.{{GPURenderPipelineDescriptor/vertexState}}.{{GPUVertexStateDescriptor/vertexBuffers}}.length:
```
Comment (kvark) [inline spec/index.bs:5656]: we need an issue or spec to say that the vertex buffers must have enough members for the draw

File: spec/index.bs [hunk 6]
```diff
@@ -5587,6 +5642,41 @@ enum GPUStoreOp {
...
+
+<div algorithm>
+    To determine if it's <dfn abstract-op>valid to draw indexed</dfn> with {{GPURenderEncoderBase}} |encoder|
+    run the following steps:
+
+    If any of the following conditions are unsatisfied, return `false`:
+        <div class=validusage>
+            - It must be [$valid to draw$] with |encoder|.
+
+            - |encoder|.{{GPURenderEncoderBase/[[index_buffer]]}} must not be `null`.
```
Comment (kvark) [inline spec/index.bs:5671]: and same for index buffers
